[dtensor][5/N] add cached propagator for TP

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* #92931
* #90735
* __->__ #90734
* #90733

This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

Differential Revision: [D42876249](https://our.internmc.facebook.com/intern/diff/D42876249)

wanchaol committed:
[dtensor][4/N] add cached propagator for TP

This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

XilunWu reviewed (approved):
thx for enabling `_CachingPropagator` in TP api.

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876249).

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

Differential Revision: [D42876249](https://our.internmc.facebook.com/intern/diff/D42876249)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876249).

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

Differential Revision: [D42876249](https://our.internmc.facebook.com/intern/diff/D42876249)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][5/N] add cached propagator for TP"


This PR adds a cached propagator for TP use, it caches the sharding
prop decision for the same input sharding on an operator. This could
improve eager mode performance.

Differential Revision: [D42876249](https://our.internmc.facebook.com/intern/diff/D42876249)

[ghstack-poisoned]

fduwjj commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -26,6 +28,9 @@
 
+# switch the DTensor propagator to use the caching propagator to speed up
+# the TP eager execution time.
+DTensor._propagator = _CachingPropagator(DTensor._propagator.op_to_rules)
```
Can we have a flag to switch this on/off. Or we want to enable it by default?

wanchaol replied:
I feel we should turn this on by default, otherwise eager mode perf would be very bad, especially once we implement more complicated sharding prop. This caching logic is relatively safe as it only change the one with the exact same input specs.

fduwjj reviewed (approved):
Thanks for making this happens! Just a small QQ. Unblock for now.

seemethere commented:
> @wanchaol has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876249).

How are these linked to the incorrect diff that was landed? From what I understand the correct diff for this is [D42888015](https://www.internalfb.com/diff/D42888015)

bigfootjon commented:
Ping me on workchat and I'll take a look tomorrow
